zeroex/orderwatch/order_watcher.go

```diff
@@ -919,82 +921,87 @@ func (w *Watcher) add(orderInfo *ordervalidator.AcceptedOrderInfo, validationBlo
+			LastUpdated:              now,
 			FillableTakerAssetAmount: orderInfo.FillableTakerAssetAmount,
-			EndState:                 zeroex.ESOrderAdded,
+			IsRemoved:                false,
+			IsPinned:                 pinned,
 		}
-		orderEvents = append(orderEvents, addedEvent)
-		stoppedWatchingEvent := &zeroex.OrderEvent{
-			Timestamp:                now,
-			OrderHash:                orderInfo.OrderHash,
-			SignedOrder:              orderInfo.SignedOrder,
-			FillableTakerAssetAmount: orderInfo.FillableTakerAssetAmount,
-			EndState:                 zeroex.ESStoppedWatching,
+		// Final expiration time check before inserting the order. We might have just
+		// changed max expiration time above.
+		if !pinned && orderInfo.SignedOrder.ExpirationTimeSeconds.Cmp(w.maxExpirationTime) == 1 {
+			// HACK(albrow): This is technically not the ideal way to respond to this
+			// situation, but it is a lot easier to implement for the time being. In the
+			// future, we should return an error and then react to that error
+			// differently depending on whether the order was received via RPC or from a
+			// peer. In the former case, we should return an RPC error response
+			// indicating that the order was not in fact added. In the latter case, we
+			// should effectively no-op, neither penalizing the peer or emitting any
+			// order events. For now, we respond by emitting an ADDED event immediately
+			// followed by a STOPPED_WATCHING event. If this order was submitted via
+			// RPC, the RPC client will see a response that indicates the order was
+			// successfully added, and then it will look like we immediately stopped
+			// watching it. This is not too far off from what really happened but is
+			// slightly inefficient.
+			addedEvent := &zeroex.OrderEvent{
+				Timestamp:                now,
+				OrderHash:                orderInfo.OrderHash,
+				SignedOrder:              orderInfo.SignedOrder,
+				FillableTakerAssetAmount: orderInfo.FillableTakerAssetAmount,
+				EndState:                 zeroex.ESOrderAdded,
+			}
+			orderEvents = append(orderEvents, addedEvent)
+			stoppedWatchingEvent := &zeroex.OrderEvent{
+				Timestamp:                now,
+				OrderHash:                orderInfo.OrderHash,
+				SignedOrder:              orderInfo.SignedOrder,
+				FillableTakerAssetAmount: orderInfo.FillableTakerAssetAmount,
+				EndState:                 zeroex.ESStoppedWatching,
+			}
+			orderEvents = append(orderEvents, stoppedWatchingEvent)
+		} else {
+			err = txn.Insert(order)
+			if err != nil {
+				if _, ok := err.(db.AlreadyExistsError); ok {
+					// If we're already watching the order, that's fine in this case. Don't
+					// return an error.
+					return orderEvents, nil
+				}
+				if _, ok := err.(db.ConflictingOperationsError); ok {
+					logger.WithFields(logger.Fields{
+						"error": err.Error(),
+						"order": order,
+					}).Error("Failed to insert order into DB")
+					return orderEvents, nil
```

Comment:
@albrow I may have missed something on your last PR, but I thought we were going to return an error here.

Reply:
Sorry I made that change in a different spot here: https://github.com/0xProject/0x-mesh/pull/756/commits/536804815b797f5e63d0598b9e4f9a9e615e0542. We should probably return an error everywhere we log one in this function, but I don't think it should block the PR.

Reply:
Yep, I agree. I saw the other changes and thought that might have happened. I can include these changes in my post-9.2.0 cleanup PR